psydac/api/fem.py

```diff
@@ -653,14 +667,37 @@ def allocate_matrices(self, backend=None):
 
         if is_broken:
             i,j = self.get_space_indices_from_target(domain, target )
+            test_fem_space   = self.spaces[1].spaces[i]
+            trial_fem_space  = self.spaces[0].spaces[j]
             test_space  = test_space.spaces[i]
             trial_space = trial_space.spaces[j]
+            ncells = tuple(max(i,j) for i,j in zip(test_fem_space.ncells, trial_fem_space.ncells))
+            is_conformal = tuple(test_fem_space.ncells) == ncells and tuple(trial_fem_space.ncells) == ncells
+
+        else:
+            ncells = tuple(max(i,j) for i,j in zip(test_fem_space.ncells, trial_fem_space.ncells))
 
         if isinstance(expr, (ImmutableDenseMatrix, Matrix)): # case of system of equations
 
             if is_broken: #multi patch
                 if not self._matrix[i,j]:
-                    self._matrix[i,j] = BlockMatrix(trial_space, test_space)
+                    mat = BlockMatrix(trial_fem_space._refined_space[ncells].vector_space, test_fem_space._refined_space[ncells].vector_space)
+                    if not is_conformal and not i==j:
+                        if all(trn>=tn for trn,tn in zip(trial_fem_space.ncells, test_fem_space.ncells)):
```

Comment:
If I understand correctly, this if condition determines whether we need a restriction or prolongation operator. I would like to make this explicit by defining the boolean variables `use_restriction` and `use_prolongation` after `is_broken` and `is_conformal`.

This approach would also allow us to raise the proper errors when the spaces are not fully compatible. E.g. when the list `[trn - tn for trn, tn in zip(trial_fem_space.ncells, test_fem_space.ncells)]` contains both strictly positive and strictly negative values.

CC: @campospinto @FrederikSchnack

Reply:
done in [86f53a3](https://github.com/pyccel/psydac/pull/241/commits/86f53a3e46503da7dac62c58b2edeedc4f631418)

Reply:
Should you not raise an error as I suggested above?

Reply:
Never mind. This can be improved in a future PR.